### PR TITLE
Eclass task 1: clean up old code

### LINKED ISSUE #71013: Remove equivalence cache from the scheduler code base
**What would you like to be added**:
Remove equivalence cache from the scheduler code base.

**Why is this needed**:
The equivalence cache (eCache) was added to the scheduler as a mechanism to improve performance of running predicate functions. The equivalence cache stores the results of predicates for pods and as long as conditions of a node are not changed, it uses the cached results for pods which have the same scheduling requirements.
While on paper, this should have improved performance of the scheduler significantly, in practice it [slowed the scheduler down](https://github.com/kubernetes/kubernetes/issues/63784) for many common scenarios. The reason turned out to be lock contention, and accessing a three-level cache which was sometime slower than running the predicates themselves.
We then tried to [optimize the locking mechanism](https://github.com/kubernetes/kubernetes/pull/65714). It helped improve performance over the previous implementation, but it was still [causing slow down](https://github.com/kubernetes/kubernetes/pull/65714#issuecomment-410016382) compared to the scheduler without eCache, for pods that didn't have complex scheduling requirements. It [improved performance for pods with inter-pod affinity/anti-affinity](https://github.com/kubernetes/kubernetes/pull/65714#issuecomment-410102800) though. This was before we added further optimizations that improved performance of affinity/anti-affinity by 5x. So, performance improvements for affinity/anti-affinity is much smaller now, but it is still significant enough to consider having eCache. However, it turned out that eCache complicates our code base much and invalidating the eCache at various events turns out to be error prone and makes building some of the new scheduling features harder. For example, ensuring that dynamic volume binding works with eCache proved to be non-trivial. As a result we have decided to remove the current implementation of eCache.

Our plan is to redesign the equivalence cache with a different mechanism to ensure that the scheduler does not keep retrying scheduling a large number of equivalent pods after it finds one of them unschedulable. When one pod is determined unschedulable, all other equivalent pods will be unschedulable as well. So, the scheduler can save CPU cycles and try other pods.

/kind cleanup

/sig scheduling

### PR DESCRIPTION
**Pull request by resouer**
*Long live equivalence class!*

Since [new design of eclass is out](https://docs.google.com/document/d/1x907Ow183mHVdmIgBO_BjZA0DPjPnB1lzCRnbBtm1Ks/edit) , it's now safe to remove the old eclass related code. Then PRs of new design can follow up.

**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
/kind cleanup

**What this PR does / why we need it**:

Task 1 of #68720 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #71013

**Special notes for your reviewer**:

cc @denkensk @Huang-Wei @bsalamat 

Marked as WIP since we need to make sure the clean up is complete and tests are happy. Will be ready for review after that.

**Does this PR introduce a user-facing change?**:

```release-note
Clean up old eclass code
```

### CODE REVIEW

**Comment by denkensk** [comment]
lgtm

**Review by Huang-Wei — APPROVED** [review]
/lgtm

**Comment by resouer** [comment]
Thanks! We still need someone from sig-testing to review. Kindly ping @jberkus, any bandwidth?

The deleted e2e tests are not enabled by default, so I think it's quite safe. :-)

**Comment by bsalamat** [comment]
@msau42 FYI

**Review by bsalamat — APPROVED** [review]
/lgtm
/approve

Thanks, @resouer! There is a small conflict that should be resolved.

**Comment by msau42** [comment]
lgtm!

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 1]
```diff
@@ -49,13 +49,6 @@ func init() {
 	registerAlgorithmProvider(defaultPredicates(), defaultPriorities())
 
 	// IMPORTANT NOTES for predicate developers:
```
Comment (misterikkit) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: I think this line should be removed also.

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 2]
```diff
@@ -1510,167 +1506,3 @@ func (c *syncingMockCache) UpdateNodeNameToInfoMap(infoMap map[string]*scheduler
 	})
 	return err
 }
-
-// TestCacheInvalidationRace tests that equivalence cache invalidation is correctly
-// handled when an invalidation event happens early in a scheduling cycle. Specifically, the event
-// occurs after schedulercache is snapshotted and before equivalence cache lock is acquired.
-func TestCacheInvalidationRace(t *testing.T) {
```
Comment (misterikkit) [inline pkg/scheduler/core/generic_scheduler_test.go]: I enjoyed writing this test. Can you also delete `syncingMockCache`?

File: pkg/scheduler/factory/factory.go [hunk 3]
```diff
@@ -365,14 +353,12 @@ func NewConfigFactory(args *ConfigFactoryArgs) Configurator {
 		},
 	)
 
-	// On add and delete of PVs, it will affect equivalence cache items
-	// related to persistent volume
+	// On add and delete of PVs.
```
Comment (misterikkit) [inline pkg/scheduler/factory/factory.go]: I think the remaining comment doesn't say anything. Let's just remove it.

File: pkg/scheduler/factory/factory.go [hunk 4]
```diff
@@ -884,24 +691,7 @@ func (c *configFactory) deletePodFromSchedulingQueue(obj interface{}) {
 }
 
 func (c *configFactory) invalidateCachedPredicatesOnUpdatePod(newPod *v1.Pod, oldPod *v1.Pod) {
```
Comment (misterikkit) [inline pkg/scheduler/factory/factory.go]: Is there a reason this function isn't deleted?

**Review by misterikkit — COMMENTED** [review]
Change looks good. Are we planning to keep the config field to enable/disable ecache?

File: pkg/scheduler/core/generic_scheduler_test.go (hunk 2 above)
Comment (resouer) [reply pkg/scheduler/core/generic_scheduler_test.go]: Haha, will do

File: pkg/scheduler/factory/factory.go (hunk 4 above)
Comment (resouer) [reply pkg/scheduler/factory/factory.go]: Emm, I will blame the latest rebase :-)

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 1 above)
Comment (resouer) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: Emm ... this line looks like a valid warning, for the below lines.

**Comment by bsalamat** [comment]
Thanks, Harry. Please take a look at the "govet" failure. It needs a small fix.

/approve

**Comment by resouer** [comment]
@bsalamat The PR is rebased frequently these days, maybe you wanna do a final check. Also, we need someone to approve the change in test, I did ping someone in comment above but seems not work ...

**Comment by misterikkit** [comment]
/lgtm
change still looks good. Just needs approval from @kubernetes/sig-testing-pr-reviews

**Comment by misterikkit** [comment]
:fireworks: :champagne: 
Looking forward to the new design/implementation

**Pull request merged**